### ISSUE TITLE
Add ability to modify DALL·E 3 generated images

### ISSUE THREAD

**Issue by jeffpaul** [10up/classifai#723]
### Is your enhancement related to a problem? Please describe.

With the update in #717, we can now start to take advantage of some of the enhancements in the DALL·E 3 model.  Let's look to extend our `Generate images` modal (e.g. /wp-admin/upload.php?action=classifai-generate-image) to add in a `Modify image` action link that further accepts textarea input which is then sent back to DALL·E 3.

When we receive the originally generated image(s) back from DALL·E 3 we'll want to ensure we capture the seed ID for each as that'll be needed for the modification step.  We'll capture the desired modifications from the user via a textarea, then send that back to DALL·E 3 (e.g. "Modify image [#] with seed [seed-ID]: <user modification text here>"), and then render the newly modified image(s) in place of the first image(s).

### Designs

![modify-image](https://github.com/10up/classifai/assets/2818133/7c7552ce-8f54-436c-b476-09ecdc643aa7)

In this rough mockup, I'm recommending we first tweak our primary button and secondary action link such that the first two options on generated images are `[Insert to Post]` and then `_Import to Media Library_`.  A third option here for `_Modify image`, when clicked, would present a textarea (with "add / remove / change image by..." placeholder text to help guide the user on what to enter) and a `[Generate modifications]` button.  Clicking that button would send the image seed ID and described modifications back to DALL·E 3 and then render in place of the prior image(s) with the newly modified image(s).

### Describe alternatives you've considered

There's some consideration for what we do with the original image prompt and modification text (especially if there are multiple modifications made).  We currently show the original prompt in the initial textarea in the `Generate images` modal, so perhaps we continue to amend that text with the modification text?  In terms of the alt text that's added to any imported/inserted generated image we could attempt to include that same prompt+modification(s) text (or albeit more lazily / easily stick with the original prompt text).

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by faisal-alvi** [comment]
@jeffpaul @dkotter 👋🏻 

After reviewing the DALL·E API [documentation](https://platform.openai.com/docs/api-reference/images), I’d like to share some insights and limitations regarding the requested enhancement:  

1. **No Seed ID for Generated Images**  
   - Currently, the DALL·E API does not provide any seed ID with generated images. Even if seed IDs were available, I couldn’t find any endpoint in the documentation that allows modifications to a generated image using seed IDs.  

2. **Existing Image Modification Options**  
   - However, the API provides a way to **edit an existing image** using [inpainting](https://platform.openai.com/docs/api-reference/images/createEdit), but it has specific requirements:  
     - **The image to edit** must be a valid PNG file, less than 4MB, and square.  
     - **The mask** must define the transparent areas where changes should be applied (e.g., fully transparent areas in the mask indicate the editable parts). This mask must also be a valid PNG, less than 4MB, and have the same dimensions as the image.  
     - Along with the image and mask, the user can provide a new prompt specifying changes for the transparent areas.  

   This approach works only for partial edits (inpainting) and is relatively limited in terms of freeform modifications.

3. **Image Variations**  
   - Another available option is generating [variations](https://platform.openai.com/docs/api-reference/images/createVariation) of an image. This requires the original image to be passed as input and results in new variations **without** user-provided prompts. This is not suitable for implementing user-driven modifications since it doesn’t allow specifying changes.  

4. **Iteration Through Prompts**  
   - The last option is iterating through prompts. However, this isn’t a new feature, it’s essentially what users can already do by editing the prompt and clicking the existing generate button.

Given these limitations, it seems that what the enhancement requests (e.g., modifying images directly using seed IDs or without requiring manual upload/masks) is not currently supported by the DALL·E API.  

Would love to hear your thoughts or suggestions on how we could approach this, considering the current API constraints.

**Comment by dkotter** [comment]
In addition to the points @faisal-alvi makes, looking at the API documentation again, I think the biggest issue here is that DALL·E 3 only supports image generation, not edits or variations (these are only supported by DALL·E 2 which we no longer support as of #717). So until this is changed by OpenAI, I don't think there is any action we can take here.

That said, I do think there are additional improvements we can make to the image generation process. I've documented those in a separate issue here: https://github.com/10up/classifai/issues/440#issuecomment-2622391002 and assigned that to @faisal-alvi

**Comment by jeffpaul** [comment]
Given that it sounds like we'll need to wait for either OpenAI to add in this functionality (large apologies on my end, I somehow thought this was already feasible) or identify an alternative AI service provider that could do this (which would be a good thing as we currently only support DALLE for this featureset), perhaps we should punt this out to `Future Release`?

**Comment by dkotter** [comment]
> perhaps we should punt this out to Future Release

Yeah, not sure there's anything else we can do. Some nice changes did come out of this for the next release (#857) but that doesn't resolve the main point in this issue, which won't be feasible until OpenAI changes things

**Comment by jeffpaul** [comment]
@dkotter perhaps this is now possible with https://openai.com/index/image-generation-api/?

**Comment by dkotter** [comment]
> [@dkotter](https://github.com/dkotter) perhaps this is now possible with https://openai.com/index/image-generation-api/?

Looks like the new model supports generations and edits but still not variations.

I do think it's worth updating to that API and adding edit support. I have a WIP PR that updates to the new model: #897. Once that's done, I think a separate PR to introduce edits is a good next step.

Worth nothing this model isn't available to everyone yet (for instance, our account needs verified before we can use it, see https://help.openai.com/en/articles/10910291-api-organization-verification)

**Comment by jeffpaul** [comment]
> Worth nothing this model isn't available to everyone yet (for instance, our account needs verified before we can use it, see https://help.openai.com/en/articles/10910291-api-organization-verification)

Our org should be verified now.

**Comment by oscarssanchezz** [comment]
From my research, it seems OpenAI does not return the seed in the response yet https://platform.openai.com/docs/api-reference/images/object 

Gemini does not seem to return the seed too. Together AI it depends on the model, and Stable Diffusion seems to support seed.

Instead of using a seed to "re-create" the current most reliable way people do this is to take the image you just generated and upload it back as a reference image in the next prompt + describing the change.

Usually, it works just fine but I've had instances where it generates something that differs from the original way too much. Also, with the current proposed UI, say a user liked the original image better and wants to go back to that. There's no guarantee the model will return the original image.

@jeffpaul @dkotter can you share your thoughts?

**Comment by dkotter** [comment]
> Instead of using a seed to "re-create" the current most reliable way people do this is to take the image you just generated and upload it back as a reference image in the next prompt + describing the change.

Yes, I think this is what the OpenAI Image Edit API supports, you pass in an image and tell it what edits you want (so no image seed support). This would be a great feature for us to add if we can figure out the best way to implement this into our existing flow.

> Also, with the current proposed UI, say a user liked the original image better and wants to go back to that. There's no guarantee the model will return the original image.

This is a risk though I'm not overly concerned with it myself, though I don't think it would be too difficult to import the image into the Media Library prior to sending it off for edits so you can easily revert if desired.
